daml-assistant/exe/DA/Daml/Assistant.hs

```diff
@@ -239,12 +259,60 @@ dispatch env path args = do
     requiredIO "Failed to spawn command subprocess." $
         runProcess_ (setEnv dispatchEnv $ proc path args)
 
-displayErrors :: IO () -> IO ()
-displayErrors m = m `catches`
-    [ Handler $ \ (e :: AssistantError) -> do
-        hPutStrLn stderr (displayException e)
-        exitFailure
-    , Handler $ \ (e :: ConfigError) -> do
-        hPutStrLn stderr (displayException e)
-        exitFailure
+handleErrors :: forall t. L.Handle IO -> IO t -> IO t
+handleErrors logger m = m `catches`
+    [ Handler (go . displayException @AssistantError)
+    , Handler (go . displayException @ConfigError)
     ]
+  where
+    go :: String -> IO t
+    go err = do
+        hPutStrLn stderr err
+        L.logJson logger L.Error $ mkLogTable
+            [ ("event", "error")
+            , ("message", A.String (T.pack err))
+            ]
+        exitFailure
+
+withLogger :: DamlPath -> (L.Handle IO -> IO ()) -> IO ()
+withLogger damlPath k = do
+    let logOfInterest prio = prio `elem` [L.Telemetry, L.Warning, L.Error]
+        gcpConfig = L.GCPConfig
+            { gcpConfigTag = "assistant"
+            , gcpConfigDamlPath = Just (unwrapDamlPath damlPath)
+            }
+
+        optedOutPath = unwrapDamlPath damlPath </> ".opted_out"
+
+    isOptedOut <- doesPathExist optedOutPath
+    if isOptedOut
+        then
+            k L.makeNopHandle
+        else do
+            L.withGcpLogger gcpConfig logOfInterest L.makeNopHandle $ \gcpState logger -> do
+                L.logMetaData gcpState
+                k logger
+
+-- | Get the arguments to `daml` and anonimize all but the first.
+-- That way, the daml command doesn't get accidentally anonimized.
+anonimizeArgs :: IO [T.Text]
+anonimizeArgs = do
+    args <- getArgs
```

Comment:
might be a bit nicer if you map `T.pack` directly over `getArgs`.

Reply:
good point, thanks!